File: src/graphics/clouds_debug.cpp
#include "clouds.h"

#include "widget/debug_console.h"
#include "imgui/imgui.h"
#include "imgui/imgui_internal.h"
#include "city/city.h"

ANK_REGISTER_PROPS_ITERATOR(config_show_clouds_properties);
void config_show_clouds_properties(bool header) {
    if (header) {
        return;
    }
    bool common_open = ImGui::TreeNodeEx("Clouds", ImGuiTreeNodeFlags_None, "Clouds");
    if (common_open) {
        {
            ImGui::BeginTable("split", 2, ImGuiTableFlags_BordersOuter | ImGuiTableFlags_Resizable);

            auto &config = g_clouds.config;
            game_debug_show_property("Reset Clouds", [i=0] { g_clouds.init_cloud_images(); });
            game_debug_show_property("num_cloud_ellipses", config.num_cloud_ellipses);
            game_debug_show_property("cloud_alpha_increase", config.cloud_alpha_increase);
            game_debug_show_property("cloud_columns", config.cloud_columns);
            game_debug_show_property("cloud_rows", config.cloud_rows);
            game_debug_show_property("cloud_width", config.cloud_width);
            game_debug_show_property("cloud_height", config.cloud_height);
            game_debug_show_property("cloud_size_ratio", config.cloud_size_ratio);
            game_debug_show_property("cloud_scale", config.cloud_scale);
            game_debug_show_property("cloud_min_creation_timeout", config.cloud_min_creation_timeout);
            game_debug_show_property("cloud_max_creation_timeout", config.cloud_max_creation_timeout);
            game_debug_show_property("pause_min_frames", config.pause_min_frames);

            pcstr status_text[3] = { "INACTIVE", "CREATED", "MOVING" };
            for (int i = 0; i < g_clouds.clouds.size(); i++) {
                ImGui::PushID(bstring32("CloudTable", i).hash());

                ImGui::TableNextRow();
                ImGui::TableSetColumnIndex(0);
                ImGui::AlignTextToFramePadding();
                bool anim_open = ImGui::TreeNodeEx(bstring32("Cloud", i).c_str(), ImGuiTreeNodeFlags_None, "Cloud %d", i);
                ImGui::TableSetColumnIndex(1);

                if (anim_open) {
                    const auto &cloud = g_clouds.clouds[i];
                    e_cloud_status status = cloud.status;

                    game_debug_show_property("status ", status_text[status]);
                    game_debug_show_property("speedx ", cloud.speed.x.current_speed);
                    game_debug_show_property("speedy ", cloud.speed.y.current_speed);
                    game_debug_show_property("pos", cloud.pos);
                    game_debug_show_property("render_pos", cloud.render_pos);

                    ImGui::TreePop();
                }

                ImGui::PopID();
            }

            ImGui::EndTable();
        }

        ImGui::TreePop();
    }
}

File: src/scripts/clouds_debug.js
log_info("akhenaten: clouds_debug")

[es=event_draw_debug_properties]
function clouds_debug_draw_properties(ev) {
    if (!imgui.tree_node_ex("Clouds")) {
        return
    }

    if (imgui.button("Reset Clouds")) {
        __clouds_reset()
    }

    imgui.begin_table("Clouds", 2, imgui.table_flags_debug_props())
    imgui.property_input("num_cloud_ellipses", __clouds_config, "num_cloud_ellipses")
    imgui.property_input("cloud_alpha_increase", __clouds_config, "cloud_alpha_increase")
    imgui.property_input("cloud_columns", __clouds_config, "cloud_columns")
    imgui.property_input("cloud_rows", __clouds_config, "cloud_rows")
    imgui.property_input("cloud_width", __clouds_config, "cloud_width")
    imgui.property_input("cloud_height", __clouds_config, "cloud_height")
    imgui.property_input("cloud_size_ratio", __clouds_config, "cloud_size_ratio")
    imgui.property_input("cloud_scale", __clouds_config, "cloud_scale")
    imgui.property_input("cloud_min_creation_timeout", __clouds_config, "cloud_min_creation_timeout")
    imgui.property_input("cloud_max_creation_timeout", __clouds_config, "cloud_max_creation_timeout")
    imgui.property_input("pause_min_frames", __clouds_config, "pause_min_frames")
    imgui.end_table()

    var count = __clouds_count()
    for (var i = 0; i < count; i++) {
        if (!imgui.tree_node_ex2("Cloud" + i, "Cloud " + i)) {
            continue
        }

        imgui.begin_table("Cloud" + i, 2, imgui.table_flags_debug_props())
        imgui.property_input("status", __clouds_cloud_status(i))
        imgui.property_input("speedx", __clouds_cloud_speedx(i))
        imgui.property_input("speedy", __clouds_cloud_speedy(i))
        imgui.property_input("pos", __clouds_cloud_pos(i))
        imgui.property_input("render_pos", __clouds_cloud_render_pos(i))
        imgui.end_table()

        imgui.tree_pop()
    }

    imgui.tree_pop()
}
